### PR TITLE
Refactor WriteThread and add pipeline write option (pipeline write part 2)

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Splitting WriteThread into WriteThread and WriteThreadImpl so that I can plugin pipeline write code by reusing the interface and some of the logic from WriteThread. Also add the db option enable_pipeline_write, which when being enable, will just return Status::NotSupported for writes for now. Other than that all logic are identical as before, just refactoring. 

This patch is stacking on #2042. More details about pipeline write will be describe in the next patch.

### CODE REVIEW

File: include/rocksdb/options.h [hunk 1]
```diff
@@ -746,6 +746,8 @@ struct DBOptions {
   // Default: 16MB/s
   uint64_t delayed_write_rate = 16 * 1024U * 1024U;
 
+  bool enable_pipeline_write = false;
```
Comment (maysamyabandeh) [inline include/rocksdb/options.h]: nit: enable_pipelined_write instead of enable_pipeline_write?

Comment (maysamyabandeh) [inline include/rocksdb/options.h]: I guess in the next commit you will add comments to explain what this option means?

Comment (yiwu-arbug) [inline include/rocksdb/options.h]: Naming and comments updated!

**Comment by yiwu-arbug** [comment]
Will land this PR after #2058 get accepted.

**Comment by yiwu-arbug** [comment]
@siying I merged WriteThread and WritePipeline in #2058. Will clean up this one.

**Pull request closed without merging**